xwayland/server.c

```diff
@@ -146,7 +146,10 @@ static void server_finish_process(struct wlr_xwayland_server *server) {
 	/* We do not kill the Xwayland process, it dies to broken pipe
 	 * after we close our side of the wm/wl fds. This is more reliable
 	 * than trying to kill something that might no longer be Xwayland.
+	 * We do need to destroy xwm however, so send a signal to give an
+	 * opportunity.
 	 */
+	wlr_signal_emit_safe(&server->events.finish, NULL);
```

Comment:
Hmm. What I think happens is that we close the X11 FD when stopping the server, but I don't think we handle this properly in the XWM.

When the X11 server dies, the XWM should cleanup and destroy itself, just like any other X11 client. Right now it seems we aren't handling FD failures properly in XWM. I think we're missing something like this in `x11_event_handler`:

```c
if ((mask & WL_EVENT_HANGUP) || (mask & WL_EVENT_ERROR)) {
	// Cleanup
}
```